### PR TITLE
Adding MERGE Predicates

### PR DESCRIPTION
**Pull request by Tang8330**
Allowing merge predicate columns to be specified. We will then use it as part of the merge clause. 

For larger tables, this is beneficial because this helps reduce the amount of data required to be scanned when Transfer merges.

### CODE REVIEW

File: lib/destination/dml/merge.go [hunk 1]
```diff
@@ -227,10 +227,10 @@ func (m *MergeArgument) GetStatement() (string, error) {
 	subQuery := fmt.Sprintf("( %s )", m.SubQuery)
 	if m.DestKind == constants.BigQuery {
 		subQuery = m.SubQuery
+	}
 
-		if len(m.AdditionalEqualityStrings) > 0 {
```
Comment (Tang8330) [inline lib/destination/dml/merge.go:231]: We only had this capability for BigQuery before, so pulling this out now.

File: clients/snowflake/writes.go [hunk 2]
```diff
@@ -49,7 +50,16 @@ func (s *Store) Merge(tableData *optimization.TableData) error {
 			}
 		}
 
-		err = shared.Merge(s, tableData, s.config, types.MergeOpts{})
+		var additionalEqualityStrings []string
+		if len(tableData.TopicConfig.AdditionalMergeColumns) > 0 {
+			for _, additionalMergeCol := range tableData.TopicConfig.AdditionalMergeColumns {
+				additionalEqualityStrings = append(additionalEqualityStrings, fmt.Sprintf("c.%s = cc.%s", additionalMergeCol, additionalMergeCol))
```
Comment (Tang8330) [inline clients/snowflake/writes.go:56]: c and cc are table aliases for staging and target table respectively.

File: lib/kafkalib/topic.go [hunk 3]
```diff
@@ -47,6 +47,7 @@ type TopicConfig struct {
 	IncludeArtieUpdatedAt     bool                        `yaml:"includeArtieUpdatedAt"`
 	IncludeDatabaseUpdatedAt  bool                        `yaml:"includeDatabaseUpdatedAt"`
 	BigQueryPartitionSettings *partition.BigQuerySettings `yaml:"bigQueryPartitionSettings,omitempty"`
+	AdditionalMergeColumns    []string                    `yaml:"additionalMergeColumns,omitempty"`
```
Comment (Tang8330) [inline lib/kafkalib/topic.go:50]: When we go to Prod, we should create a separate type for this so we can embed additional information

File: clients/snowflake/writes.go [hunk 4]
```diff
@@ -49,7 +50,16 @@ func (s *Store) Merge(tableData *optimization.TableData) error {
 			}
 		}
 
-		err = shared.Merge(s, tableData, s.config, types.MergeOpts{})
+		var additionalEqualityStrings []string
+		if len(tableData.TopicConfig.AdditionalMergeColumns) > 0 {
+			for _, additionalMergeCol := range tableData.TopicConfig.AdditionalMergeColumns {
```
Comment (nathan-artie) [inline clients/snowflake/writes.go:55]: Does `additionalMergeCol` need to be escaped?

**Review by nathan-artie — APPROVED** [review]

File: clients/snowflake/writes.go (hunk 4 above)
Comment (Tang8330) [reply clients/snowflake/writes.go:55]: Yeah it does / should. Will add

**Review by danafallon — APPROVED** [review]

**Pull request merged**